stdlib/internal/promql/join.md

```diff
@@ -0,0 +1,91 @@
+### join
+
+`join` merges two input streams into a single output stream.
+Records that have the same group key and `_time` values will be joined in the output.
+The output groups will be the same as the input groups.
+
+`join` has the following properties:
+
+| Name    | Type     | Description |
+| ------- | -------- | ----------- |
+| tables  | object   | Specifies the two table streams to join. |
+| fn      | function | Defines the function that joins the rows of each table. The return value is an object which defines the output record structure. |
+
+Ex.
+```
+import "internal/promql"
+
+left = from(bucket: "my-bucket")
+    |> range(start: -1h)
+    |> filter(fn: (r) => r._field == "A")
+    |> drop(columns: ["_field"])
+    |> rename(columns: {_value: "a"})
+
+right = from(bucket: "my-bucket")
+    |> range(start: -1h)
+    |> filter(fn: (r) => r._field == "B")
+    |> drop(columns: ["_field"])
+    |> rename(columns: {_value: "b"})
+
+promql.join(tables: {left, right}, fn: (t) => {
+    return {t.left with b: t.right.b}
+}
+```
+
+`join` assumes that each input table is sorted by `_time`.
+The planner will ensure this by inserting a `sort` procedure after each one of join's immediate predecessors.
+`join` will only call `fn` on rows that pass the join condition.
```

Comment:
What is the "join condition"?  Is it just that the `_time` columns match?

Reply:
For rows that do not have a corresponding row with the same `_time` in the other table, I guess you just discard the row?  Is that what PromQL does for implicit joins?  I could imagine also doing a left/right/full outer join where all or some of the rows are retained, and using null values for missing rows.

Reply:
A row on the left will only join with a row on the right if both rows have the same group key value as well as the same `_time`  value. Everything else is discarded. And yes this is what PromQL does. PromQL doesn't support the other types of joins. Flux can support these other join types with an extra predicate parameter, but to keep things simple, for the PromQL case, I chose not to do that.

Reply:
There is an exception for the `or` (set union) and `unless` (set difference) binary operators.

`or` produces output series (Flux rows) for label sets (Flux group keys) that are present *either* on the LHS or RHS. The sample value (Flux `_value`) is taken from the LHS if the series is present there, from the RHS otherwise.

`unless` propagates anything from the LHS into the output unless there is a matching label set (Flux group key) on the RHS.

I guess this would need some type of outer join that fills up missing rows on either side with `null` values, on which one could then do conditional postprocessing to achieve those operators?

Reply:
@juliusv I did not include transpilation examples for `or` and `unless` but they can indeed be transpiled to Flux using a combination of left and right join operations, as well as union. I can add more examples to this doc

Reply:
Ah ok, if this can be done somehow, then great :) Are you generally taking over the transpilation work for vector-to-vector binops? (I'd be happy about that, just good to know.)